### PR TITLE
refactor: look up badge number by rfid per request

### PR DESCRIPTION
**Pull request by skyqrose**
Asana Task: Noticed during [📐 Orbit: allow manually associating a badge serial with an operator for testing](https://app.asana.com/0/1200689710863995/1207890347662644/f), discussed in slack.

Instead of sending all badge_serials to the frontend with the employee data, this PR makes a new api endpoint to look them up one by one during sign in.

I'll need help testing this on sandbox (but I see sandbox is in use right now).

- Tests:
  - `(x)` Has tests
  - `( )` Doesn't need tests
  - `( )` Tests deferred (with justification)
- Product/Design sign off:
  - `( )` Okayed the plan for the feature (e.g. the design files, or the Asana task)
  - `( )` Reviewed the feature as implemented (e.g. on dev-green, or saw screenshots)
  - `(x)` No review needed

### CODE REVIEW

File: lib/orbit_web/controllers/rfid_controller.ex
Comment (skyqrose) [inline lib/orbit_web/controllers/rfid_controller.ex:13]: This will crash if a badge serial isn't unique. Which is fine for now cuz we want to fix that anyway.

File: js/test/api.test.ts
Comment (lemald) [inline js/test/api.test.ts:56]: **nitpick:** this test name is a big confusing because the error isn't in `parser`, maybe this would be more descriptive?
```suggestion
  test("rejects on return data type error", async () => {
```

File: js/components/operatorSignIn/operatorSelection.tsx
Comment (lemald) [inline js/components/operatorSignIn/operatorSelection.tsx:30]: **suggestion:** I do want to flag that there's a potential race condition here where the `useEffect` dependencies change before this promise resolves. I think that because of the way the `useNfc` hook works in practice (it "locks in" the result of a successful scan) it's unlikely that it will cause an actual problem in practice but I haven't yet 100% convinced myself of that.

File: test/support/factory.ex
Comment (lemald) [inline test/support/factory.ex:19]: **question:** what's the purpose of removing this from the factory? We should still have the badge serials available on the back-end, it's just that the controller won't preload them in its query and won't include a `badge_serials` key in its response.

File: js/components/operatorSignIn/operatorSelection.tsx
Comment (skyqrose) [reply js/components/operatorSignIn/operatorSelection.tsx:30]: Yeah, I'd been thinking about that too. I was leaning towards it being unlikely enough to pop up that I just didn't bother to put in the effort to fix it. Maybe I should leave a comment nearby acknowledging that the race condition exists, even if I don't fix it now?

Comment (lemald) [reply js/components/operatorSignIn/operatorSelection.tsx:30]: A comment sounds good! I think that the way that the `useNfc` hook works is the key part of the rationale here to mention.

File: test/support/factory.ex
Comment (skyqrose) [reply test/support/factory.ex:19]: No serials seemed like a simpler default value. Usually if you care that this field exists, you should set it in the test, so that changing the factory values doesn't break tests. There was only one test that depended on this and I ended up editing it anyway.

**Review by lemald — APPROVED** [review]
There's still my [nitpick comment about test naming](https://github.com/mbta/orbit/pull/104#discussion_r1739306397) but aside from that this looks good.

File: js/test/api.test.ts
Comment (skyqrose) [reply js/test/api.test.ts:56]: I think the step that's failing could be called parsing, and zod even calls this `zod.parse()`. I think it's actually `parser` that's misnamed, since that's about transforming typed data into other typed data, rather than turning unknown data into typed data. (The name is left over from before we used zod in Glides, and that step actually did some parsing.)

**Comment by skyqrose** [comment]
I think I want to wait until Tuesday to merge this so we have a chance to do an end-to-end test of this on sandbox. (I'm not set up to test it locally.)

File: js/test/api.test.ts
Comment (lemald) [reply js/test/api.test.ts:56]: I guess I'd say that really neither should be called parsing strictly speaking because they're not about taking a string and tokenizing it and turning it into an abstract syntax tree, but if Zod uses the terminology then _I guess_ that's okay and it's not a hill I'll choose to die on. 😝

**Comment by skyqrose** [comment]
Tested this manually with @mathcolo and it works.

**Pull request merged**
